Ensure References are consistent with Proxy Object

dwil opened the pull request:
This intercepts proxied methods and attributes on the inside the Reference and updates the `_id` and `_href`
attributes to ensure they are synchronized with wrapped object.

bmacphee commented on analyzere/base_resources.py:
```diff
@@ -48,6 +52,35 @@ def __deepcopy__(self, memo):
             return copy.deepcopy(self.__wrapped__, memo)
         return Reference(self._href)
 
+    def __getattribute__(self, name):
```
gross, but if it fixes the problem in the current implementation, I guess it's legit.

DanielArndt commented on analyzere/base_resources.py:
```diff
@@ -48,6 +52,35 @@ def __deepcopy__(self, memo):
+        def update_ref(item):
+            try:
+                id_ = item.id
+                href_ = urljoin(analyzere.base_url, item._get_path(id_))
```
A `Resource` seems to be the only thing that has `_get_path` so would it make more sense for this to live there instead?

dwil replied:
That is correct. So this protects against someone just doing something like `Reference.id = <something>` and then the `_id` and `_href` getting updated to the new id and url without them called `Reference.save()` first.

DanielArndt replied:
Hmm... there seems to be some weird relationship between `Reference` and `Resource` that I'm not following here. `Reference` itself seems to just hide that it is actually going to return a `Resource`. Anyway, that's not a result of these changes. 

I guess I find it confusing that we're relying on a method that only exists in `Resource`. Why not override the `__getattribute__` of `Resource` instead? What cases would we be missing?

dwil replied:
If we were to do that on the `Resource` would that still have access to the `_id` and `_href` attributes of the `Reference`? Since those are the two properties we're trying to update.

DanielArndt replied:
Ahh, ok. Maybe that's what I was missing. I thought `Resource` was a subclass of `Reference`.

DanielArndt commented on analyzere/base_resources.py:
```diff
@@ -48,6 +52,35 @@ def __deepcopy__(self, memo):
+                return retval
+            return newfunc
+        else:
+            update_ref(Proxy)
```
Will this ever work? This passes in the class `Proxy`, but then we call `item.id` in `update_ref()`. I would assume that it hits the `try...except` here, because the class itself will not have the attribute `id` but I may be missing something. Python objects always confuse me.

dwil replied:
I think you're correct here.

dwil commented:
@DanielArndt 

Okay I think this is better. I've added in a number of tests that ensure that we're not accidentally changing class level attributes and handle special cases for class methods that should not be able to change where the reference is point to. With regards to the points above:

1. These seem like that have to be there. There is something about the `Proxy` class that requires them if you also want to use instanced versions of the attributes.

2. Removed these for calls to `self.<attr>`

3. Removed that, you're right it was unnecessary.

4. Added in a number of tests like this.

DanielArndt commented on analyzere/base_resources.py:
```diff
@@ -48,6 +53,43 @@ def __deepcopy__(self, memo):
+
+        attr = Proxy.__getattribute__(self, name)
+        if hasattr(Reference, name):
+            return attr
```
Doesn't this skip anything that is a class member of `Reference` (and either not a member of `Proxy`, or overridden)?

For example, if you get `Reference.__copy__()`, I think this will return `Proxy`'s `__copy__()`.
